src/control/lib/hardware/libfabric/bindings.go

```diff
@@ -41,13 +41,71 @@ uint lib_fabric_version(void)
 {
 	return FI_VERSION(1, 7);
 }
+
+// call into C functions that are looked up at runtime
+
+int call_fi_getinfo(void *fn, struct fi_info *hint, struct fi_info **info)
+{
+	int (*getinfo)(uint, char*, char*, ulong, struct fi_info*, struct fi_info**);
+
+	assert(fn != NULL);
+	getinfo = (int (*)(uint, char*, char*, ulong, struct fi_info*, struct fi_info**))fn;
+	return getinfo(lib_fabric_version(), NULL, NULL, 0, hint, info);
+}
+
+struct fi_info *call_fi_dupinfo(void *fn, struct fi_info *info)
+{
+	struct fi_info *(*dupinfo)(struct fi_info*);
+
+	assert(fn != NULL);
+	dupinfo = (struct fi_info *(*)(struct fi_info*))fn;
+	return dupinfo(info);
+}
+
+void call_fi_freeinfo(void *fn, struct fi_info *info)
+{
+	void (*freeinfo)(struct fi_info*);
+
+	assert(fn != NULL);
+	freeinfo = (void (*)(struct fi_info*))fn;
+	freeinfo(info);
+}
+
+char *call_fi_strerror(void *fn, int code)
+{
+	char *(*strerr)(int);
+
+	assert(fn != NULL);
+	strerr = (char *(*)(int))fn;
+	return strerr(code);
+}
 */
 import "C"
 
 import (
+	"fmt"
+	"unsafe"
+
 	"github.com/pkg/errors"
+
+	"github.com/daos-stack/daos/src/control/lib/dlopen"
 )
 
+// Load dynamically loads the libfabric library and provides a method to unload it.
+func Load() (func(), error) {
+	hdl, err := openLib()
+	if err != nil {
+		return nil, err
+	}
+	return func() {
+		hdl.Close()
+	}, nil
+}
+
+func openLib() (*dlopen.LibHandle, error) {
+	return dlopen.GetHandle([]string{"libfabric.so.1", "libfabric.so"})
```

Comment:
This is probably fine for the foreseeable future, but I'm curious as to how adding support for multiple versions of libfabric (or ucx, or whatever) might change this design.

Reply:
My intention wasn't to bind too tightly to a specific version, since we'll fall back to "libfabric.so" if "libfabric.so.1" isn't available. Perhaps having the .1 isn't that meaningful at all in the long run, but we're requesting API version 1.7 from the library when we make the fi_getinfo call. My understanding is that libfabric handles some API versioning issues internally.